HW2/P3/mandelbrot.pyx

```diff
@@ -16,56 +40,96 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
                  int max_iterations=511):
     cdef:
        int i, j, iter
-       np.complex64_t c, z
+       int n_threads  # added to allow setting of number of threads
 
-       # To declare AVX.float8 variables, use:
-       # cdef:
-       #     AVX.float8 v1, v2, v3
-       #
-       # And then, for example, to multiply them
-       #     v3 = AVX.mul(v1, v2)
-       #
-       # You may find the numpy.real() and numpy.imag() fuctions helpful.
+       np.float32_t [:, :] r_coords, i_coords  # store real and imag parts
+       int n_8cols  # number of 8 column chunks
+
+       AVX.float8 cr, ci, zr, zi  # c and z 
+       AVX.float8 iter_ind  # iteration counter for individual pixels
+       AVX.float8 magz, newzr, newzi  # temporary values associated with z
+       AVX.float8 mask, tmp, new_iter_ind
 
     assert in_coords.shape[1] % 8 == 0, "Input array must have 8N columns"
     assert in_coords.shape[0] == out_counts.shape[0], "Input and output arrays must be the same size"
     assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
 
+    # separate real and imaginary components
+    r_coords = np.real(in_coords)
+    i_coords = np.imag(in_coords)
+
     with nogil:
-        for i in range(in_coords.shape[0]):
-            for j in range(in_coords.shape[1]):
-                c = in_coords[i, j]
-                z = 0
+
+        n_threads = 4
+        n_8cols = in_coords.shape[1] / 8  # break into chunks of 8 columns
+
+        for i in prange(in_coords.shape[0], num_threads=n_threads,
+                        schedule='static', chunksize=1):
+
+            for j in range(n_8cols):
+
+                cr = AVX.make_float8(r_coords[i, j*8+7],
+                                     r_coords[i, j*8+6],
+                                     r_coords[i, j*8+5],
+                                     r_coords[i, j*8+4],
+                                     r_coords[i, j*8+3],
+                                     r_coords[i, j*8+2],
+                                     r_coords[i, j*8+1],
+                                     r_coords[i, j*8])
+
+                ci = AVX.make_float8(i_coords[i, j*8+7],
+                                     i_coords[i, j*8+6],
+                                     i_coords[i, j*8+5],
+                                     i_coords[i, j*8+4],
+                                     i_coords[i, j*8+3],
+                                     i_coords[i, j*8+2],
+                                     i_coords[i, j*8+1],
+                                     i_coords[i, j*8])
+
```

Comment:
You could avoid the multiplication by using the optional `step` argument to [`range`](https://docs.python.org/2/library/functions.html#range), e.g.:
```
for j in range(0, n_8cols, 8)
```
```